components/widgets/inputs/users_emails_input.jsx

```diff
@@ -0,0 +1,201 @@
+            defaultMessage: this.props.loadingMessageDefault || 'Loading',
+        });
+    }
+
+    getOptionValue = (user) => {
+        return user.id || user.value;
+    }
+
+    formatUserName = (user) => {
+        let displayName = '@' + user.username + ' - ' + user.first_name + ' ' + user.last_name;
+        if (user.nickname) {
+            displayName = displayName + ' (' + user.nickname + ')';
+        }
+        return displayName;
+    }
+
+    formatShortUserName = (user) => {
+        if (user.first_name || user.last_name) {
+            return user.first_name + ' ' + user.last_name;
+        }
+        return user.username;
+    }
+
+    formatOptionLabel = (user, options) => {
+        const profileImg = imageURLForUser(user);
+        const avatar = (
+            <img
+                className='avatar'
+                alt={`${user.username || 'user'} profile image`}
+                src={profileImg}
+            />
+        );
+
+        if (options.context === 'menu') {
+            if (user.value && isEmail(user.value)) {
+                return this.getCreateLabel(user.value);
+            }
+            return (
+                <React.Fragment>
+                    {avatar}
+                    {this.formatUserName(user)}
+                </React.Fragment>
+            );
+        }
+
+        if (user.value && isEmail(user.value)) {
+            return (
+                <React.Fragment>
+                    <MailIcon className='mail-icon'/>
+                    <span>{user.value}</span>
+                </React.Fragment>
+            );
+        }
+
+        return (
+            <React.Fragment>
+                {avatar}
+                {this.formatShortUserName(user)}
+            </React.Fragment>
+        );
+    }
+
+    onChange = (value) => {
+        if (this.props.onChange) {
+            this.props.onChange(value.map((v) => {
+                if (v.id) {
+                    return v;
+                }
+                return v.value;
+            }));
+        }
+    }
+
+    getCreateLabel = (value) => (
+        <React.Fragment>
+            <MailPlusIcon className='mail-plus-icon'/>
+            <FormattedMarkdownMessage
+                key='widgets.users_emails_input.valid_email'
+                id={this.props.validAddressMessageId || t('widgets.users_emails_input.valid_email')}
+                defaultMessage={this.props.validAddressMessageDefault || 'Add **{email}**'}
+                values={{email: value}}
+            />
+        </React.Fragment>
+    );
+
+    NoOptionsMessage = (props) => {
+        const inputValue = props.selectProps.inputValue;
+        if (inputValue) {
+            return (
+                <div className='users-emails-input__option'>
+                    <FormattedMarkdownMessage
+                        id={this.props.noMatchMessageId || t('widgets.users_emails_input.no_user_found_matching')}
```

Comment:
Here we could extract `id` and `defaultMessage` to variables conditioned by `inputValue` and avoid having the whole component twice, right?

Reply:
I'll do it